vm/utils.c

```diff
@@ -0,0 +1,24 @@
+#include <stdio.h>
+#include <sys/stat.h>
+
+#include "tr.h"
+#include "internal.h"
+
+char *read_file_into_string(char *filename, int *str_len) {
+  FILE *fp;
+  struct stat stats;
+  
+  if (stat(filename, &stats) == -1) return NULL;
+  fp = fopen(filename, "rb");
+  if (!fp) return NULL;
+  
+  char *string = TR_ALLOC_N(char, stats.st_size + 1);
+  size_t bytes_read = fread(string, 1, stats.st_size, fp);
+  fclose(fp);
+  if (bytes_read == (size_t)stats.st_size) {
+    string[stats.st_size] = 0;
+    *str_len = stats.st_size;
+    return string;
+  }
```

Comment:
in the else case here, you'd leak the memory of string

Reply:
I thought it will be recycled by the garbage collector, is my understanding wrong?

Reply:
yes, but imo we should free stuff that we ourselves use (and thus control, unlike user's scripts) anyway. that improves code quality (code can reused elsewhere) and decreases pressure on the GC.

Reply:
ok, I agree, I'll update this PR shortly